Add owo-lib rich translations compatibility

kikugie opened the pull request:
Rich translations allow creating text with configurable effects using owo-lib: https://docs.wispforest.io/owo/rich-translations/  
This format translates nicely to yaml:
```yml
soundboard:
  title:
    - text: Soundboard
      color: aqua
      bold: true
 ```
 In the plugin it is a simple change - whenever the endpoint value is a list, it preserves all its contents and converts to json as is.

kikugie committed:
allow lists in translations

Fallen-Breath commented:
Suggestion: create a boolean switch for it, since it's not that a Minecraft standard. Benefits:

- People won't trigger this feature unintentionally
- Better implemetation control if there are similiar features in the future

kikugie committed:
the amount of hacks needed is insane

Gradle, **why** is it so difficult to pass parameters to a filter? You already have it for extensions, just do it here again!

kikugie commented:
![image](https://github.com/Fallen-Breath/yamlang/assets/50152507/8a9d7929-d302-4565-a1cd-9eef2525d83b)

Fallen-Breath commented:
Are these complex implementations really necessary? thonk

kikugie commented:
Apparently yes, because the filtering reader doesn't accept additional constructor args, like extensions do, so the alternative is either to create a non-static inner class or an anonymous class to get variables from the outer scope. The third way messes with how gradle injects passed properties via reflection, but that requires implementing non-api interfaces.

kikugie commented:
Any updates on this? Or preferred other methods?

Fallen-Breath commented:
> Any updates on this? Or preferred other methods?

I want to have a try myself to see if there's any better implementation on this when I have time. If you really needs this feature, you can use your fork for now (e.g. via jitpack)

kikugie committed:
Inner class approach

kikugie commented:
A cleaner approach, where `Yamlang2JsonlangMapper` is an inner class that can pull fields from `YamlangConvertor`

Fallen-Breath commented:
Yeah this approach is much nicer

Fallen-Breath commented on src/main/java/me/fallenbreath/yamlang/YamlangConvertor.java:
```diff
@@ -71,4 +87,54 @@ private static String renameYaml2Json(String fileName)
 		return baseName + JSON_PREFIX;
 	}
+
+	private String yamlang2Jsonlang(String ymlContent)
```
I think keeping `yamlang2Jsonlang` and `parseMap` inside the `Yamlang2JsonlangMapper` will be better, cuz these 2 methdods are only used in `Yamlang2JsonlangMapper`

kikugie replied:
They can't be because
1. Non-static inner classes can't have static methods in Java 8.
2. Those methods can't be members of `Yamlang2JsonlangMapper` because they are called before `super`.

Fallen-Breath commented on src/main/java/me/fallenbreath/yamlang/YamlangConvertor.java:
```diff
@@ -71,4 +87,54 @@ private static String renameYaml2Json(String fileName)
+
+	public class Yamlang2JsonlangMapper extends FilterReader
+	{
+		public Yamlang2JsonlangMapper(Reader reader) throws IOException {
```
`{` in new line

Fallen-Breath commented on src/main/java/me/fallenbreath/yamlang/YamlangExtension.java:
```diff
@@ -50,4 +50,14 @@ public interface YamlangExtension
+	 * Default value: false
+	 * @see <a href="https://docs.wispforest.io/owo/rich-translations/">OWO-LIB Rich Translations</a>
+	 */
+	Property<Boolean> getRichTranslations();
```
The optional name "rich translations" is confusing. I don't think it's specific to OWO-LIB. I think a name like `owolibRichTranslations` that explictly show that it's a feature from OWO-LIB is better

Same for the `ConverterConfiguration`

Fallen-Breath replied:
Also same change requested on the comment of this method, like

```diff
+Enables <a href="https://docs.wispforest.io/owo/rich-translations/">OWO-LIB Rich Translations</a> in the yaml translation files
-Enables the rich translation support in yaml files
```

Fallen-Breath replied:
tldr don't use vague names like `richTranslations` or `allowLists`, use names like `owolibRichTranslations` that explictly shows it's the rich translation feature from owolib

Fallen-Breath commented on src/main/java/me/fallenbreath/yamlang/ConverterConfiguration.java:
```diff
@@ -0,0 +1,10 @@
+
+public class ConverterConfiguration
+{
+    public final boolean allowLists;
```
tab, no space, owo

kikugie committed:
Requested cleanups

kikugie commented:
Hhhh big bummer: 
![image](https://github.com/Fallen-Breath/yamlang/assets/50152507/8ac01158-bf94-4ac0-ad3c-765468f5fa99)

Gradle requires a constructor that only accepts a `Reader`, where as inner classes sneakily get another argument.

kikugie commented:
Since Gradle is not very welcoming of configurable filters, this is going to be cursed anyway. I'd leave as default behavior because you don't use lists in normal translation files.  
The last alternative is to throw away the Gradle copy spec function and simply do the same process manually.  
If no other solution works, I'll just host the fork on my maven in case someone needs the rich translation support.

Fallen-Breath committed:
cleaner impl

Fallen-Breath commented:
You can simply use this method to pass argument to the filter: https://docs.gradle.org/current/javadoc/org/gradle/api/file/CopySpec.html#filter(java.util.Map,java.lang.Class)

Changes are already pushed to your branch and it's now working. You can have another check to see if that's what you want
